File: binding/src/webchannel/libqwebchannel.cpp
#include <QObject>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <type_traits>
#include <QWebChannel>
#include <QWebChannelAbstractTransport>
#include <qwebchannel.h>
#include "libqwebchannel.h"
#include "libqwebchannel.hxx"

QWebChannel* QWebChannel_new() {
	 return new QWebChannel();
}

QWebChannel* QWebChannel_new2(QObject* parent) {
	 return new QWebChannel(parent);
}

libqt_string QWebChannel_Tr(const char* s) {
	QString _ret = QWebChannel::tr(s);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QWebChannel_RegisterObjects(QWebChannel* self, const libqt_map objects) {
	self->registerObjects(QHash<QString, QObject *>());
}

libqt_map QWebChannel_RegisteredObjects(const QWebChannel* self) {
	auto _ret = self->registeredObjects();
	libqt_map _map;
	_map.len = _ret.size();
	_map.keys = nullptr;
	_map.values = nullptr;
	return _map;
}

void QWebChannel_RegisterObject(QWebChannel* self, const libqt_string id, QObject* object) {
	self->registerObject(QString::fromUtf8(id.data, id.len), object);
}

void QWebChannel_DeregisterObject(QWebChannel* self, QObject* object) {
	self->deregisterObject(object);
}

bool QWebChannel_BlockUpdates(const QWebChannel* self) {
	return self->blockUpdates();
}

void QWebChannel_SetBlockUpdates(QWebChannel* self, bool block) {
	self->setBlockUpdates(block);
}

int QWebChannel_PropertyUpdateInterval(const QWebChannel* self) {
	return self->propertyUpdateInterval();
}

void QWebChannel_SetPropertyUpdateInterval(QWebChannel* self, int ms) {
	self->setPropertyUpdateInterval(ms);
}

void QWebChannel_BlockUpdatesChanged(QWebChannel* self, bool block) {
	self->blockUpdatesChanged(block);
}

void QWebChannel_Connect_BlockUpdatesChanged(QWebChannel* self, intptr_t slot) {
    void (*slotFunc)(QWebChannel*, bool) = reinterpret_cast<void (*)(QWebChannel*, bool)>(slot);
    QWebChannel::connect(self, &QWebChannel::blockUpdatesChanged, [self, slotFunc](bool block) {
	slotFunc(self, block);
    });
}

void QWebChannel_ConnectTo(QWebChannel* self, QWebChannelAbstractTransport* transport) {
	self->connectTo(transport);
}

void QWebChannel_DisconnectFrom(QWebChannel* self, QWebChannelAbstractTransport* transport) {
	self->disconnectFrom(transport);
}

libqt_string QWebChannel_Tr2(const char* s, const char* c) {
	QString _ret = QWebChannel::tr(s, c);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QWebChannel_Tr3(const char* s, const char* c, int n) {
	QString _ret = QWebChannel::tr(s, c, n);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QWebChannel_Delete(QWebChannel* self) {
    delete self;
}

File: binding/include/webchannel/libqwebchannel.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qwebchannel.html
pub const qwebchannel = struct {

    /// New constructs a new QWebChannel object.
    pub fn New() QtC.QWebChannel {
        return qtc.QWebChannel_new();
    }


    /// New2 constructs a new QWebChannel object.
    pub fn New2(parent: ?*anyopaque) QtC.QWebChannel {
        return qtc.QWebChannel_new2(@ptrCast(parent));
    }


    pub fn Tr(s: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QWebChannel_Tr(@ptrCast(s));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qwebchannel.Tr: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn RegisterObjects(self: ?*anyopaque, objects: *const QHash<QString, QObject *>) void {
        qtc.QWebChannel_RegisterObjects(@ptrCast(self), @ptrCast(objects));
    }

    pub fn RegisteredObjects(self: ?*anyopaque, ) QHash<QString, QObject *> {
        return qtc.QWebChannel_RegisteredObjects(@ptrCast(self));
    }

    pub fn RegisterObject(self: ?*anyopaque, id: []const u8, object: ?*anyopaque) void {
        const id_str = qtc.libqt_string{
    .len = id.len,
    .data = id.ptr,
};
qtc.QWebChannel_RegisterObject(@ptrCast(self), id_str, @ptrCast(object));
    }

    pub fn DeregisterObject(self: ?*anyopaque, object: ?*anyopaque) void {
        qtc.QWebChannel_DeregisterObject(@ptrCast(self), @ptrCast(object));
    }

    pub fn BlockUpdates(self: ?*anyopaque, ) bool {
        return qtc.QWebChannel_BlockUpdates(@ptrCast(self));
    }

    pub fn SetBlockUpdates(self: ?*anyopaque, block: bool) void {
        qtc.QWebChannel_SetBlockUpdates(@ptrCast(self), block);
    }

    pub fn PropertyUpdateInterval(self: ?*anyopaque, ) i32 {
        return qtc.QWebChannel_PropertyUpdateInterval(@ptrCast(self));
    }

    pub fn SetPropertyUpdateInterval(self: ?*anyopaque, ms: i32) void {
        qtc.QWebChannel_SetPropertyUpdateInterval(@ptrCast(self), ms);
    }

    pub fn BlockUpdatesChanged(self: ?*anyopaque, block: bool) void {
        qtc.QWebChannel_BlockUpdatesChanged(@ptrCast(self), block);
    }

    pub fn OnBlockUpdatesChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque, bool) callconv(.c) void) void {
        qtc.QWebChannel_Connect_BlockUpdatesChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn ConnectTo(self: ?*anyopaque, transport: ?*anyopaque) void {
        qtc.QWebChannel_ConnectTo(@ptrCast(self), @ptrCast(transport));
    }

    pub fn DisconnectFrom(self: ?*anyopaque, transport: ?*anyopaque) void {
        qtc.QWebChannel_DisconnectFrom(@ptrCast(self), @ptrCast(transport));
    }

    pub fn Tr2(s: []const u8, c: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QWebChannel_Tr2(@ptrCast(s), @ptrCast(c));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qwebchannel.Tr2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Tr3(s: []const u8, c: []const u8, n: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QWebChannel_Tr3(@ptrCast(s), @ptrCast(c), n);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qwebchannel.Tr3: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QWebChannel_Delete(@ptrCast(self));
    }
};
